lib/uplink/list_test.go

```diff
@@ -0,0 +1,233 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink_test
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"storj.io/storj/internal/testcontext"
+	"storj.io/storj/internal/testplanet"
+	"storj.io/storj/lib/uplink"
+	"storj.io/storj/pkg/storj"
+)
+
+type expectedResult struct {
+	path     string
+	isPrefix bool
+}
+
+type listChallenge struct {
+	commands        []string         // list commands that should result in the same set of responses
+	expectedResults []expectedResult // results that should come back for each command above
+}
+
+type putGetListTest struct {
+	bucket         string          // bucket to upload to
+	paths          []string        // test will create a file at every path here
+	listChallenges []listChallenge // set of list commands and expected results
+}
+
+// TestPutGetList allows you to create a bucket, a set of paths, and a set of
+// list challenges that you want to try against that configuring. It will create
+// a unique file for each path, upload it, download it, and run the list
+// challenges against it.
+func TestPutGetList(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount:   1,
+		StorageNodeCount: 1,
+		UplinkCount:      1},
+		func(t *testing.T, ctx *testcontext.Context, planet *testplanet.Planet) {
+			apiKey := planet.Uplinks[0].APIKey[planet.Satellites[0].ID()]
+			satelliteAddr := planet.Satellites[0].Local().Address.Address
+
+			tests := []putGetListTest{
+				{
+					bucket: "bu1",
+					paths: []string{
+						"a-file",
+						"a/b-file",
+						"a/b/slash-file",
+						"a/b////",
+						"a/b//",
+						"a/b//c-file",
+						"a/b//c/",
+						"//bob",
+						"/",
+					},
+					listChallenges: []listChallenge{
```

Comment:
A couple of these challenges only have the same results for all the commands for encrypted paths, right? Specifically, `[]string{"a", "a/"}`, and `[]string{"a/b", "a/b/"}`.